wikimedia/wikipedia-ios
# (WIP) Add per Wikipedia language localizations to support grammatically correct translations.

**Phabricator ticket:** https://phabricator.wikimedia.org/T231566

### Notes
Still a little bit left to do, but wanted to get this up so we can all look at it. The "high impact" tag is more about how verbose this is rather than it being high touch. And the question tag is for us all to decide: is a verbose fix like this actually worth it?

* Adds `localized_language_strings` script to generate extension to `CommonStrings.swift` (manual integration for now)
* Replaces these duplicated localized strings with `localized_language_strings` output
```
"explore-featured-article-sub-heading-from-wikipedia" = "From Wikipedia";
"explore-nearby-placeholder-sub-heading-on-wikipedia" = "From Wikipedia";
"explore-random-article-sub-heading-from-wikipedia" = "From Wikipedia";
"in-the-news-sub-title-from-wikipedia" = "From Wikipedia";

"explore-featured-article-sub-heading-from-language-wikipedia" = "From %1$@ Wikipedia";
"explore-nearby-placeholder-sub-heading-on-language-wikipedia" = "From %1$@ Wikipedia";
"explore-random-article-sub-heading-from-language-wikipedia" = "From %1$@ Wikipedia";
"in-the-news-sub-title-from-language-wikipedia" = "From %1$@ Wikipedia";

"explore-most-read-sub-heading-on-wikipedia" = "On Wikipedia";
"explore-most-read-sub-heading-on-language-wikipedia" = "On %1$@ Wikipedia";
"talk-page-info-active-conversations" = "Active conversations on %1$@ Wikipedia";
"top-read-header-with-language" = "%1$@ Wikipedia";
```
* Uses new strings in explore feed, talk pages, and top read widget. 

I'm not very fond of this approach, but in terms of producing the most accurate and grammatically correct translations, I don't believe we landed on anything better yet. Instead of having translators just translate language adjectives in isolation, this approach allows them to translate the entire string in _the_ most contextually relevant to the language way. This comes at the cost of a _lot_ of strings to translate, so there's definitely going to be some lift for kind and patient translators. I am not a language expert, and instead of requiring any of us to be language experts and understand the classifications of which languages grammatically do what, we can now allow native language translators to do what's best for these particular strings.

If by some insight someone has a less verbose solution, I'm happy to discuss!

### Test Steps
- Add in app languages other than English
- Explore feed subheadings
	- Confirm these strings render with the site-specific language inserted
		- `From (LANGUAGE) Wikipedia`
		- `On (LANGUAGE) Wikipedia`
- Talk Pages 
	- Head into a user talk page
	- Confirm `Active conversations on (LANGUAGE) Wikipedia` displays with the site-specific language inserted
- Top read widget
	- Confirm the expanded state header displays `(LANGUAGE) Wikipedia` with the site-specific language inserted


### My TODOs: 
- [ ] Add documentation to localization.md
- [ ] Convert `"explore-nearby-sub-heading-your-location-from-language-wikipedia" = "Your location from %1$@ Wikipedia";`
- [ ] Poke around with the generic fallback strings - I'm not sure if I'm getting the desired behavior yet
- [ ] Add more detail to Notes and Test cases when I remove the WIP
